keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,469 @@
+
+Take a look at the example:
+
+- `curReplicas = 10`
+- `calculatedReplicas = 20`
+
+The user specifies only one HPA parameter `constraints.scaleUp.pods = 5` and expects that number of replicas to be set to 15 during the next HPA controller loop.
+But the algorithm picks the largest change instead:
+
+    Constraint1 = 10 * 2 = 20 (as constraints.scaleUp.percent = 100 by default)
+    Constraint2 = 10 + 5 = 15 (as constraints.scaleUp.pods = 5, set by the user)
+    scaleUpLimit = max(20, 15) = 20
+    desiredReplicas = 20
+
+The user might expect that the autoscaler would use the smallest constraint (15), not the largest one (20). This behavior is not intuitive, but it does make sense if considered thoroughly.
+
+The main idea of the HPA is to autoscale because of a load increase to avoid request failures. It should work on both small clusters and large ones. For small clusters, the absolute number constraint works better (ScaleUpPods), for large clusters the percentage works better (ScaleUpPercentage).
+
+Example: If the current cluster size is `1` and calculated cluster size for this particular load is `20`, then we should reach it ASAP.
+
+For default values (ScaleUpPercent = 100, ScaleUpPods = 4) and “pick the largest constraint” concept, we’ll increase 1 -> 20 in 3 steps
+
+    1 -> 5 -> 10 -> 20
+
+The first step will use “scaleUp.pods” constraint; the next steps will use “scaleUp.percent” one.
+
+In case of more intuitive “pick the hardest limit” concept, we’ll increase the cluster in 6 steps:
+
+    1 -> 2 -> 4 -> 8 -> 12 -> 16 -> 20
+
+Given that each step takes [90 sec in worst case], we’ll respond to the load increase in `(6-3)*90 sec = ~ 5 min`.
+
+That’s too much, we should respond faster than that.
+
+[90 sec in worst case]: https://dzone.com/articles/kubernetes-autoscaling-101-cluster-autoscaler-hori-1
+
+The scale down constraints are a method to prevent too rapid loss of capacity. Hence, it makes sense to pick the maximum of two constraints.
+
+#### Stabilization Window
+
+Currently stabilization window ([PR][], [RFC][], [Algorithm Details][]) is used to gather “scale-down-recommendations” during some time (default is 5min),
+and a new number of replicas is set to the maximum of all recommendations.
+
+It may be defined by command line option `--horizontal-pod-autoscaler-downscale-stabilization-window`.
+
+[PR]: https://github.com/kubernetes/kubernetes/pull/68122
+[RFC]: https://docs.google.com/document/d/1IdG3sqgCEaRV3urPLA29IDudCufD89RYCohfBPNeWIM/edit#heading=h.3tdw2jxiu42f
+[Algorithm Details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### API Changes
+
+The following API changes are needed:
+
+We should add `Scale{Up,Down}{Percent,Pods}` fields into the HPA spec
+
+The resulting solution will look like this:
+
+```golang
+type HPAScaleConstraintValue struct {
+    Rate         *HPAScaleConstraintRateValue
+    DelaySeconds *int32
```

Comment:
if I understand, this is not a "delay" as much as a "window" ?  Please think about the name that best captures the semantics.

Also document bounds.  Can I set a delay of 14 days?  How much buffer are we willing to allocate (and lose if the controller crashes) ?